> **Jupyter slideshow:** This notebook can be displayed as slides. To view it as a slideshow in your browser, run the following cell:

> `> jupyter nbconvert [this_notebook.ipynb] --to slides --post serve`
 
> To toggle off the slideshow cell formatting, click the `CellToolbar` button, then `View > Cell Toolbar > None`.

In [1]:
# ! jupyter nbconvert Intro_to_Python.ipynb --to slides --post serve

## Did you install Anaconda?

- Install Anaconda (https://www.anaconda.com/distribution/)

> Make sure to install the latest (python 3.7) version 

<img src="./images/anaconda.png" width="600" height="600" align="center"/>

- We will go over the Python Programming language in next session

<img src="./images/salesforce.svg" width="50" height="50" align="right"/>

<img src="./images/di.png" width="50" height="50" align="right"/>

# Intro to Python 

<img src="./images/python.jpg" width="300" height="300" align="center"/>


### Why Python?

<img src="./images/whypython.png" width="700" height="500" align="center"/>

<a id="learning-objectives"></a>
### Learning Objectives
*In this lesson, we will go over the following:*

- Data types in Python.
- Functions in Python.
- Control flows
- Examples Programs.

### Lesson Guide

- [Into to jupyter (10–15 min)](#survey1)
- [Common Python Types (15 min)](#types-def)
- [Common Types Code-Along (20 min)](#types-codealong)
- [Python Control Flows (20 min)](#functions-def)
- [Common Python Functions Code-Along (20 min)](#functions-codealong)
- [Recap and Requests (5–10 min)](#recap-requests)


<a id="types-def"></a>
## Common Python Types

---

**In Python, what do you understand a "type" to be?**

### What is a type?

> There are several _standard_ data types within Python (and typically any programming language, although it varies as to what is or isn't in a type for each). 

> A type tells the computer what can or can't be done with a given piece of information in Python (or whichever programming language you are using). For example, we can **add** and **subtract** numbers and we can **count** the number of letters in a word. Rather than try (and fail!) to add two words together or count the number of numbers in a number, types let the computer know efficiently what can and cannot be done in a given situation.

<img src="./images/datatypes.png" width="600" height="600" align="center"/>

### Types in Python


There are two basic groups of types in Python — single elements (one item or piece of information) and collections (groups of things, which can be either single elements themselves, other collections, or a mix of both).

**Single Elements**

- **Integers:** Whole numbers ranging from negative infinity to infinity, such as 1, 0, -5, etc.
- **Floats:** Short for "floating point number;" usually used with decimals, such as 2.8 or 3.14159.
- **Strings:** A set of letters, numbers, or other characters, e.g., "The fox is quick." — any set of values that contains a non-numeric character.

**Collections**

- **Tuples:** An ordered sequence with a fixed number of elements; e.g., in `x = (1, 2, 3)`, the parentheses makes it a tuple. `x = ("Kirk", "Picard", "Spock")` — once you've defined this, you can't change it.
- **Lists:** An ordered sequence without a fixed number of elements, e.g., `x = [1, 2, 3]`. Note the square brackets. `x = ["Lord", "of", "the", "Rings"]` — this can be changed as you like.
- **Dictionaries**: An unordered collection of key-value pairs, e.g., `x = {'Mark': 'Twain', 'Apples': 5}`. To retrieve each value (the part after each colon), use its key (the part before each colon). For example, `x['Apples']` retrieves the value `5`.

**In your own words, what is the difference between a list and a dictionary? Can you think of any real-life examples of lists or dictionaries?**

### list vs. tuple vs. dictionary 

Typically, lists and dictionaries are most popular because of their ease of use and applicability to numerous situations (by contrast, because tuples are immutable/cannot be modified; they are used more sparingly). Lists have an inherent order (the first element is at index `0`, the second element is at index `1`, etc.) and we can call each element by that ordinal number (such as `x[0]` or `x[100]`). Dictionaries do not have an order (so `x[0]` will fail), but they use the name of the key to return that element.

**Example**: Think about the difference between a sign-up list and an address book:
- **Sign-up list**: We would refer to the order of who signed up when (i.e., "the eighth person to sign up").
- **Address book**: We would refer to a person by looking for their name (i.e., "the contact info for Bill Personson").

### Variables

Variables are names that have been assigned to specific values or data. These names can be almost anything you want, but there are some restrictions and best practices.


#### Best Practices
- Usually people use _ instead of spaces in a variable name

my_number = 10

>**Instructor Note:** List their rules on the whiteboard and then bring up some of the concepts below.

#### Restrictions

- Variable names cannot be just a number (i.e., 2, 0.01, 10000).
- Variables cannot be assigned the same name as a default or imported function (i.e., 'type', 'print', 'for').
- Variable names cannot contain spaces.

#### Best Practices

- Variable names should be lowercase.
- A variable's name should be representative of the value(s) it has been assigned. For example, instead of `a` or `money`, something like `money_usd` tells the user both what is stored in that variable *and* what units it's in.

<a id="types-codealong"></a>
## Common Types  (Code-Along)


<img src="./images/hands_on.jpg" width="100" height="100" align="center"/>

---

In this section, we'll practice establishing some types and common practices. To run each cell, use `shift+enter` or the `run` button in the Jupyter Notebook toolbar.

Some common reasons why we need to keep types in mind:

1) Different types can lead to different results: <br>
     - 1 + 1 results in 2, while '1' + '1' results in the string 11.
2) Operations may not work with specific types: <br>
     - len('a word') will return the number of characters in a word, while `len(25)` will return an error because numbers do not have a length.

In [6]:
# Type The above question answer here 


In [12]:
# Assigning a float:
x = 1.0
type(x)

float

In [13]:
# Assigning an int:
y = 1
type(y)

int

In [14]:
# Assigning a string:
z = '1'
type(z)

str

**Remember that, when we're assigning variables, we are not stating that "x equals 1," we are stating that "x has been assigned the value of 1."**

### Operators

Operators can be used in a mathematical sense to calculate (or create) the sum, difference, product, or quotient of values or variables.

Note that `print()` below will print out the values of whatever is inside of the parentheses.

<img src="./images/operators.png" width="500" height="500" align="center"/>

In [81]:
# Addition:
print(1 + 2)
# Subtraction:
print(1 - 2)
# Multiplication:
print(1 * 2)
# Division:
print(1 / 2)

3
-1
2
0.5


 There is also `//` division, whose output will be the rounded-down whole number.

In [82]:
# Division of float numbers:
print(3.0 // 2)
print(-3.0 // 2)

1.0
-2.0


In [9]:
# Exponent power operator:
2 ** 3

8

In [10]:
# The modulo operator can be used to get the remainder — what's left over after the term has been cleanly divided:
# 7= 3*2 + 1 
7%2

1

### Booleans and Boolean Evaluation Operators

>**Instructor Note:** Guide students to the following definition:

Booleans exist as either true or false and are generally used as a means of evaluation.

#### Using Booleans

Booleans are frequently used to filter data or conditions. Sometimes, we may want all countries with populations greater than 4,000,000 or all people named Bob. Both of these result in a `True` or `False` condition that split our data into the groups we want.

In Python, there are several built-in commands for deciding how to filter results:

- `and`: Are both A and B true?
- `not`: Is A the same as B?
- `or`: Is A or B true?

In [8]:
True and False

False

In [13]:
not False

True

In [14]:
True or False

True

**Comparisons**

<img src="./images/comp.png" width="500" height="500" align="center"/>

In [88]:
2 > 1, 2 < 1, 2 > 2, 2 < 2, 2 >= 2, 2 <= 2

(True, False, False, False, True, True)

In [6]:
# Equality:
[1,2] == [1,2]

True

In [7]:
[1,2] == [1,2] or [2,2] == [2,1]

True

## Hands_on Challenge

#### Now You Try!
<img src="./images/hands_on.jpg" width="100" height="100" align="right"/>


With a partner, create three comparisons using `!=`, `>=`, and `<`.

In [12]:
## Type your answer here



### Strings

**What are strings? How would we use them? Can you think of any examples of strings?**

>**Instructor Note:** Guide your students to the following definition:

Strings are essentially any character combination in between quotes. They are most often used as a way of storing text. Strings are used frequently, because most of the data that humans create are text-based, such as restaurant reviews or emails.

In [18]:
s = "Hello world"
type(s)

str

In [19]:
ss='Hi World!'
type(ss)

str

Strings have a lot of associated methods and attributes that allow us to better understand and manipulate them.

**In your own words, why would we want to manipulate or change strings?**

>**Instructor Note:** Some examples include:
- Fixing misspelled words.
- Changing casing (upper, lower).
- Looking for specific words.

In [10]:
# Finding the length of the string:
print(s)
len(s)

Hello world


11

In [18]:
# Replacing an element of a string:
s2 = s.replace("world", "test")
print(s2)

Hello test


### String Indexing

In some cases, we may want a part of the string (like the first character for alphabetizing or categorizing). Indexing helps us do that.

We can extract characters at specific index locations in a string using indexing.

### In Python we use [ ] for Indexing (for lists, tuples, disctionaries, ...)

<img src="./images/indexing.jpeg" width="500" height="500" align="center"/>

In [13]:
test_string='Monty Python'

test_string[0:3]

'Mon'

In [21]:
# Indexing the first (index 0) character in the string:
print(s)
s[0]

Hello world


'H'

The number you enter after the variable name in brackets (the `[0]`) is called the index (its plural is indices).

_Counting in Python and many other programming languages begins at zero, as opposed to one. This is called zero-based indexing._

In [96]:
# This is called "splicing." We start at the left index 
#   and go up to but not include the right index.

# Objects at indexes 0, 1, and 2:
s[0:3]

'Hel'

## Indexing

[start_index : end_index : step_size]

Most ranges, or functions with ranges, have upper ends that are not inclusive. So, a range of `[0:5]` starts at `0` and stops before `5`.

A good mental trick is to look at something like `[5:25]` and say out loud "Starting at five and going up to (but not including) 25."

In [11]:
# From index 6 up to the end of the string:
print(s)
s[6:]

Hello world


'world'

In [98]:
# No start or end specified:
s[:]

'Hello world'

In [99]:
# Can we index from the right side?
s[-1]

'd'

In addition to specifying a range, you can include a step size or character skip rate. This might be helpful if you want every other letter, for example. 

These indexing methods can also be used on lists, where asking for every other number might be a good use case.

In [22]:
# Every second character starting at 0 and ending at 10:
print(s)
s[0:10:2]

Hello world


'Hlowr'

In [101]:
# Define a step size of 2; i.e., every other character:
s[::2]

'Hlowrd'

In [12]:
# The same, but for a list of numbers:
[0, 1, 2, 3, 4, 5, 6][::2]

[0, 2, 4, 6]

### Concatenating

**In your own words, what is concatenating? When might you use it?**

To add two strings together, type the first string, a `+` sign, and then the second string.

> **Instructor Note:** One easy example is adding a greeting to someone's name: `'Hello ' + name`.

In [13]:
x = 'Hello '
y = 'world'

x + y

'Hello world'

In [24]:
# Conversion from int to str is required!

dice_roll = 3

#print('You rolled a ' + dice_roll + '.')  
print('You rolled a ' + str(dice_roll) + '.')  

You rolled a 3.


#### Now You Try!
<img src="./images/hands_on.jpg" width="100" height="100" align="right"/>

Create your own string of at least 12 characters or more and:
1) Test to make sure that it is at least 12 characters long. <br>
2) Replace all of the vowels with the string `vowel`. <br>
3) Use concatenation to add another string to it. <br>

In [3]:
# Type your answer here

my_string= 'This is the first time Im writing python code'



## Lists

**What are some examples of lists? What do you remember from before?**

Lists can be composed of ints, floats, strings, or other lists, as well as other data types we haven't covered yet.

In [17]:
m = [1, 2, 3, 4]

print(type(m))
print(m)

<class 'list'>
[1, 2, 3, 4]


In [18]:
# The a variable's contents can be reassigned to another variable:
a = m

In [19]:
print(a)

[1, 2, 3, 4]


In [30]:
# List of strings:
names = ['Carol', 'Anne', 'Jessica']
print(names)

['Carol', 'Anne', 'Jessica']


### Methods

Many types have what are known as "methods:" built-in functionality that allows them to do certain things. We've already seen a couple, such as the `.replace()` method, which lets you replace words in strings. 

Lists also have several methods that allow us to alter them, such as the `.append()` method, which allows us to add another element to the end of a list.

In [31]:
names.append('Michelle')
names

['Carol', 'Anne', 'Jessica', 'Michelle']

Lists can indexed the same way strings — this allows us to target a specific value or range of values in a list without having to create a new one.

In [32]:
print(names[1:3])
print(names[::2])   # Increments the index by 2 each time (skips alternate elements).

['Anne', 'Jessica']
['Carol', 'Jessica']


In [33]:
names

['Carol', 'Anne', 'Jessica', 'Michelle']

In [34]:
names[1]

'Anne'

In [28]:
# We can slice a value in a list as well:
names[1][1:]

'nne'

Note that we always read indexing from left to right. In the example above, the interpreter looks up names and gets the first element, which is the string `"Anne"`. Then, the slice (`[1:]`) adds the first index of that string to the end of the original string, evaluating to `"nne"`.

Interestingly, the following works in the same way. Instead of having to look up the value of names, the list is directly specified (just read the line from left to right!).

In [35]:
['Carol', 'Anne', 'Jessica', 'Michelle'][1][1:]

'nne'

In [30]:
# Lists don't have to be the same type:
l = [1, 'a', 1.0, 1-1j]
print(l)

[1, 'a', 1.0, (1-1j)]


In [31]:
# We can create a list of values in a range using the range() function:
start = 10
stop = 30
step = 2
print(type(range(start, stop, step)))

# range() produces a "generator," which is beyond the scope of this introduction!
# It is often convenient to have the generator 
#    generate all of its values by converting it to a list:
list(range(start, stop, step))

<class 'range'>


[10, 12, 14, 16, 18, 20, 22, 24, 26, 28]

In [32]:
list(range(1,10,2))

[1, 3, 5, 7, 9]

Use the `.insert()` method to add values at specific indices.

In [33]:
names

['Carol', 'Anne', 'Jessica', 'Michelle']

In [34]:
names.insert(2, 'Ellen')
names

['Carol', 'Anne', 'Ellen', 'Jessica', 'Michelle']

The `.remove()` method can be used to remove specific values if they appear in a list.

In [35]:
names.append('Jeremy')
print(names)
names.remove('Jeremy')
print(names)

['Carol', 'Anne', 'Ellen', 'Jessica', 'Michelle', 'Jeremy']
['Carol', 'Anne', 'Ellen', 'Jessica', 'Michelle']


#### Now You Try!

<img src="./images/hands_on.jpg" width="100" height="100" align="right"/>


Create a list of five elements and do the following:

1) Print the last three elements. <br>
2) Insert two new elements at index 2 and append one element to the end. <br>
3) Remove one element of your choice. <br>
4) Print every other element in your list. <br>

In [11]:
##### Type your answer here
mn=[1,3,4,'hello',[4,5,7]]
mn[2:]
mn.insert(2,'insert here')
del mn[2]
print(mn)

[1, 3, 4, 'hello', [4, 5, 7]]


## Tuples

Tuples are similar to lists in that they store a sequence of various separate values. However, tuples are not mutable in that, once they are created, their values cannot be changed.

**In your own words, why would creating something that cannot be changed later be helpful?**

>**Instructor Note:** Guide students to think about safety — what if you absolutely needed to make sure that one or two numbers or values stayed the same?

In [31]:
point = (10, 20)
print(point)
print(type(point))

(10, 20)
<class 'tuple'>


In [32]:
# They can be sliced, just like lists and strings:
point[0]

10

In [33]:
point[0]=23

TypeError: 'tuple' object does not support item assignment

## Dictionaries

Dictionaries are a non-ordered Python data type. Instead of using an ordered index to access data stored in a dictionary, we use a system of key-value pairs.

**In your own words, why would we use this when we could just use a list?**

- A key is similar to a variable name.
- A value is similar to the value assigned to the variable.
- Curly braces (`{ }`) enclose dictionaries. The first input in a dictionary pair is the "key." The second input in a dictionary pair is the "value." Remember to make `key:value` pairs!

The general format looks like this:

>**Instructor Note:** Some answers might include:
- When there are lots of elements and we don't want to remember which element number is which.
- When we don't care about the order in which things happen.
     - As opposed to a list, which is kind of like a queue.

In [24]:
params = {'key1' : 1.0,
          'key2' : 2.0,
          'key3' : 3.0,}

print(type(params))
print(params)

<class 'dict'>
{'key1': 1.0, 'key2': 2.0, 'key3': 3.0}


The keys stay the same, but the values are changeable. You can also only have one occurrence of a key in a dictionary, but you can have all of the values be the same.

In [25]:
# Value for parameter2 in the params dictionary:
params['key2']

2.0

In [26]:
# Adding a new dictionary entry:
params['key4'] = 'D'

In [27]:
print(params)

{'key1': 1.0, 'key2': 2.0, 'key3': 3.0, 'key4': 'D'}


In [28]:
# Reassigning the value of a key-value pair in the dictionary:
params['key1'] = 'A'
params['key2'] = 'B'

In [29]:
print(params)

{'key1': 'A', 'key2': 'B', 'key3': 3.0, 'key4': 'D'}


In [30]:
print('Key 1 = ' + str(params['key1']))
print('Key 2 = ' + str(params['key2']))
print('Key 3 = ' + str(params['key3']))
print('Key 4 = ' + str(params['key4']))

Key 1 = A
Key 2 = B
Key 3 = 3.0
Key 4 = D


In [31]:
params.items()

dict_items([('key1', 'A'), ('key2', 'B'), ('key3', 3.0), ('key4', 'D')])

In [32]:
# Dictionaries also have methods.

# Convert a dictionary to a list of tuples (key-value pairs).
# This is later used to conveniently loop through a dictionary:
list(params.items())

[('key1', 'A'), ('key2', 'B'), ('key3', 3.0), ('key4', 'D')]

In [35]:
params.keys()

dict_keys(['key1', 'key2', 'key3', 'key4'])

In [36]:
params.values()

dict_values(['A', 'B', 3.0, 'D'])

<a id="types-quiz"></a>
## Practice: Types

<img src="./images/hands_on.jpg" width="100" height="100" align="right"/>



---

_Identify the variable types of the following five items:_

- `1`
- `-1.0`
- `1,000,000`
- `'10'`
- `('twenty-four')`


### Challenge 
_Create a list of all numbers between `1` and `100`, inclusive, using the `range()` function discussed above. Can you slice the list so that we see every fifth number, starting at `4` and ending at `82`?_

In [1]:
### Type your answer here



<a id="functions-def"></a>
## Common Python Functions and Control Flow
---


<img src="./images/control.png" width="700" height="700" align="center"/>

---
In this section, we're going to tackle some common design patterns in Python. The first is the concept of control flow — this is how our programs will return different results based on specific input. Second, we'll cover basic functions — these let us create snippets of code that we can call later in a script, which creates code that's easier to read and maintain. Remember, we're going to be reading code much more often than writing it!

## `if… else` Statements

In Python, indentation matters! This is especially true when we look at the control structures in this lesson. In each case, a block of indented code is only run some of the time. There will always be a condition in the line preceding the indented block that determines whether the indented code is run or skipped.

### `if` Statement
The simplest example of a control structure is the `if` statement. We start with `if`, followed by something that can evaluate to `True` or `False` (such as any of the comparison operators we discussed earlier).

In [29]:
if 1 == 1:
    print('The integer 1 is equal to the integer 1.')
    print('Is the next indented line run, too?')

The integer 1 is equal to the integer 1.
Is the next indented line run, too?


In [12]:
if 'one' == 'two':
    print("The string 'one' is equal to the string 'two'.")

print('---')
print('These two lines are not indented, so they are always run next.')

Notice that, in Python, the line before every indented block must end with a colon (`:`). In fact, it turns out that the `if` statement has a very specific syntax:

```
if <expression>:
    <one or more indented lines>
```

When the `if` statement is run, the expression is evaluated to `True` or `False` by applying the built-in `bool()` function. If the expression evaluates to `True`, the code block is run; otherwise, it is skipped.

#### Now You Try!
<img src="./images/hands_on.jpg" width="100" height="100" align="right"/>


Create your own string called `test_string`, then fill in the blanks here to create an `if... else` statement for whether or not the first character in `test_string` is a lowercase `a`.

In [22]:
#### Type your answer here
test_string='dapple is a company' 
    

#### `if` ... `else`

In many cases, you may want to run some code if the expression evaluates to `True` and some other code if it evaluates to `False`. This is done using `else`. Note how it is at the same indentation level as the `if` statement, followed by a colon, followed by a code block. Let's see it in action.

In [43]:
if 50 < 30:
    print("50 < 30.")
else:
    print("50 >= 30.")
    print("The else code block was run instead of the first block.")

print('---')
print('These two lines are not indented, so they are always run next.')

50 >= 30.
The else code block was run instead of the first block.
---
These two lines are not indented, so they are always run next.


#### `if` ... `elif` ... `else`

Sometimes, you might want to run one specific code block out of several. For example, perhaps we provide the user with three choices and want something different to happen with each one.

`elif` stands for `else if`. It belongs on a line between the initial `if` statement and an (optional) `else`. 

In [43]:
health = 55

if health > 70 :
    print('You are in great health!')
elif health > 40:
    print('Your health is average.') 
    print('Exercise and eat healthily!')
elif health > 20:
    print('Your health is ok.')
    print('Dont eat meat!')
else:
    print('Your health is low.')
    print('Please see a doctor now.')

print('---')
print('These two lines are not indented, so they are always run next.')

Your health is average.
Exercise and eat healthily!
---
These two lines are not indented, so they are always run next.


This code works by evaluating each condition in order. If a condition evaluates to `True`, the rest are skipped.

**Let's walk through the code.** First, we let `health = 55`. We move to the next line at the same indentation level — the `if`. We evaluate `health > 70` to be `False`, so its code block is skipped. Next, the interpreter moves to the next line at the same outer indentation level, which happens to be the `elif`. It evaluates its expression, `health > 40`, to be `True`, so its code block is run. Now, because a code block was run, the rest of the `if` statement is skipped.

## `for` Loops
---
<img src="./images/for.png" width="500" height="500" align="right"/>



One of the primary purposes of using a programming language is to automate repetitive tasks. One example is the `for` loop.

The `for` loop allows you to perform a task repeatedly on every element within an object, such as every name in a list.


Let's see how the pseudocode works:

```python
# For each individual object in the list
    # perform task_A on said object.
    # Once task_A has been completed, move to next object in the list.
```

Let's say we wanted to print each of the names in the list, as well as "is Awesome!" In this case, we'd create a temporary variable for each element in the collection (`for name in names` would put each name, in sequence, under the temporary variable `name`) and then do something with it.

In [23]:
names = ['Rebecca Bunch', 'Paula Proctor', 'Heather Davis']

for item in names:
    print(item + ' Is Awesome!')

Rebecca Bunch Is Awesome!
Paula Proctor Is Awesome!
Heather Davis Is Awesome!


We can also combine `if... else` statements and `for` loops:

In [60]:
names

['Rebecca Bunch', 'Paula Proctor', 'Heather Davis']

In [59]:
for item in names:
    if item == 'Paula Proctor':
        print(item + ' Is REALLY AWESOME!')
    else:
        print(item + ' Is Awesome!')

Rebecca Bunch Is Awesome!
Paula Proctor Is REALLY AWESOME!
Heather Davis Is Awesome!


#### Now You Try!
<img src="./images/hands_on.jpg" width="100" height="100" align="right"/>
Create a new `if... elif... else` and `for` loop combination, using a list of your own choice. 

In [26]:
### Type your answer here

nums=[1,23,4,6,7,76, 77,35]
# only print the even numbers

        


## Functions
---

**When would you want to call the same code over and over again? What benefit does that have in programming?**



>**Instructor Note:** Guide students to think about how using functions can make programming easier and ensure that we will only need to make changes/define options once in a script instead of multiple times.

Similar to the way we can use `for` loops as a means of performing repetitive tasks on a series of objects, we can also create functions to perform repetitive tasks. Within a function, we can write a large block of action and then call the function whenever we want to use it.  


Let's write some pseudocode, which is code that Python will not run successfully, but illustrates the basic idea without worrying about correct syntax:
```python
# Define the function name and the requirements it needs.
    # Perform actions.
    # Optional: Return output.
```

A function is defined like this:

```python
def function_name(arguments):
    # Do things here.
    return value
```

We start with `def` and the name of our function, then a set of parentheses. The terms we put in the parentheses will be passed into the function and stored in those variables. Finally, if we want to store the results of the function, we use `return`, which will let us take some value and store it once the function has run, like this:

```python
x = function_name(20)
```

Whatever follows `return` when the function is defined will be passed out of the function and stored in `x`.

Let's create a function that takes two numbers as arguments and returns their sum, difference, and product. 

In [51]:
def arithmetic(num1, num2):
    print(num1 + num2)
    print(num1 - num2)
    print(num1 * num2)

    
arithmetic(3,5)

8
-2
15


<a id="functions-codealong"></a>
## Common Functions Code-Along


<img src="./images/hands_on.jpg" width="100" height="100" align="right"/>
---

In this section, we'll run through some basic functions and how we might use them.

>**Instructor Note:** In this section, we'll run through some basic function generation so students can see the process of how to create and use functions. It can be helpful to walk through the code and ask students what each line does. A good suggestion would be to delete the latter sections of code and create them for your students in person so that they can see your thought process. 

Write a function that takes the length of a side of a square as an argument and returns its area.

In [14]:
def area_square(length):
    return length**2

print(area_square(4))

16


Write a function that takes the height and width of a triangle and returns its area.

In [50]:
def area_triangle(height, width):
    return height * (0.5*width)

print(area_triangle(2, 6))

6.0


Write a function that takes a string as an argument and returns a tuple consisting of two elements:

- A list of all of the characters in the string.
- A count of the number of characters in the string.

In [62]:
def list_and_count(word):
    list_of_characters = []
    for char in word:
        list_of_characters.append(char)
    return list_of_characters, len(word)

print(list_and_count('Lisa Simpson'))

(['L', 'i', 's', 'a', ' ', 'S', 'i', 'm', 'p', 's', 'o', 'n'], 12)


#### Now You Try!

Write a function that takes two integers, passed as strings, and returns the sum, difference, and product as a tuple (with all values as integers).

In [55]:
#### Type your answer here



In [56]:
def integerify(string1, string2):
    int1 = int(string1)
    int2 = int(string2)
    val_sum = int1 + int2
    val_diff = int1 - int2
    val_prod = int1 * int2
    return val_sum, val_diff, val_prod

integerify('20', '100')

(120, -80, 2000)

Write a function that takes a list as the argument and returns a tuple consisting of two elements:

- A list with the items in reverse order.
- A list of the items in the original list that have an odd index.

> **Instructor Note:** This function is a good example for introducing a couple of more advanced topics and could be written together with students versus walking them through it as it appears.

In [144]:
names

['Rebecca Bunch', 'Paula Proctor', 'Heather Davis']

In [145]:
def reverse_and_odd(input_list):
    reversed_list = input_list[::-1]
    odd_indices = []
    for i in range(len(input_list)):
        if i % 2 == 1:
            odd_indices.append(input_list[i])
    return reversed_list, odd_indices

reverse_and_odd(names)

(['Heather Davis', 'Paula Proctor', 'Rebecca Bunch'], ['Paula Proctor'])

_Write a function that calculates the factorial of an input number_

4 ! = 4 * 3 * 2 * 1      ==> 24

5 ! = 5 * 4 * 3 * 2 * 1       ==> 120

1 != 1

<img src="./images/hands_on.jpg" width="100" height="100" align="right"/>

---



In [3]:
def factorial (input):
    # add your code here
    output=1   
    
    return output
        
    
factorial (4)   

1

## Different Approaches to Solving this problem,

In [59]:
def factorial (input):
    if input == 1:
        return(1)
    else:
        return(input * factorial (input-1))
        
    
factorial (4)   

24

In [67]:
## Write your code here
def factorial (input_number):
    result=1
    for item in range(1,input_number+1):
        result=result*item
    return(result)

factorial(4)

24

In [60]:
def factor(x):
    fact = x
    while x > 1:
        fact = fact * (x-1)
        x = x-1
    return fact
print(factor(4))

24


## Take Home Practice: Functions

<img src="./images/hands_on.jpg" width="100" height="100" align="right"/>

---

Can you tackle these three challenges on your own?

_Write a function that takes a word as an argument and returns the number of vowels in the word._

_Write a function that takes in a list of animals. Have it print out each animal's name in FULL CAPITAL LETTERS._ 

- **Note:** You may need to do some outside research to find out how Python can capitalize all letters in a string! 

## Next : we will learn how to manipulate Data in Python 